### PR TITLE
Apply masking before softmax

### PR DESCRIPTION
**Pull request by adamantivm**
Adds a hack to deal with the case when the masked policy ends up being all zeroes. Now instead, we return a unifirm distribution over the valid actions.

### CODE REVIEW

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py [hunk 1]
```diff
@@ -45,12 +45,16 @@ def evaluate(self, game: Quoridor):
...
-        # Re-normalize probabilities after masking.
-        policy_probs = policy_masked
-        policy_probs = policy_probs / policy_probs.sum()
+        if np.all(policy_masked == 0):
+            print("Policy is all zeros")
+            print(policy_masked)
+            print(game)
+
+        # Apply softmax to convert to probabilities
+        policy_probs = F.softmax(torch.from_numpy(policy_masked), dim=0).numpy()
```
Comment (jonbinney) [inline deep_quoridor/src/agents/alphazero/nn_evaluator.py:57]: Hmm... if you take the softmax here, then you need to take the logarithm of the training data you use for the network, right? Otherwise the network is training on probabilities, but you're treating its output as log probabilities.

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py [hunk 2]
```diff
@@ -45,12 +45,16 @@ def evaluate(self, game: Quoridor):
         # Mask the policy to ignore invalid actions
         valid_actions = game.get_valid_actions()
         valid_action_indices = [self.action_encoder.action_to_index(action) for action in valid_actions]
-        policy_masked = np.zeros_like(unmasked_policy)
+        policy_masked = np.full_like(unmasked_policy, -1e32)
```
Comment (jonbinney) [inline deep_quoridor/src/agents/alphazero/nn_evaluator.py:48]: `-1x10**32`? Could you add a comment explaining this here?

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py [hunk 3]
```diff
@@ -45,12 +45,16 @@ def evaluate(self, game: Quoridor):
...
         policy_masked[valid_action_indices] = unmasked_policy[valid_action_indices]
 
-        # Re-normalize probabilities after masking.
-        policy_probs = policy_masked
-        policy_probs = policy_probs / policy_probs.sum()
+        if np.all(policy_masked == 0):
+            print("Policy is all zeros")
+            print(policy_masked)
+            print(game)
```
Comment (jonbinney) [inline deep_quoridor/src/agents/alphazero/nn_evaluator.py:54]: I'm not a fan of adding debug prints to committed code - it gets out of hand fast.

File: deep_quoridor/src/renderers/computation_times.py [hunk 4]
```diff
@@ -31,6 +31,8 @@ def end_arena(self, game, results: list[GameResult]):
                     f"{max(computation_times_ms):.3f}",
                     f"{statistics.median(computation_times_ms):.3f}",
                 ]
+                if num_moves > 0
```
Comment (jonbinney) [inline deep_quoridor/src/renderers/computation_times.py:34]: Thanks for fixing this! I didn't encounter the problem often since it usually crashed earlier :-)

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py (hunk 2 above)
Comment (alejandromarcu) [reply deep_quoridor/src/agents/alphazero/nn_evaluator.py:48]: Won't this give a sligth chance of taking an invalid action? probably they won't be chosen if there's any valid action with a reasonable probability, but in cases like we were debugging where the masked policies were all almost 0, it may end up chosing an invalid action.  I think what we would need to ensure is that some of the valid actions are > 0, we could take one randomly otherwise to avoid division by 0 but I think it's a red flag in the first place that the network is so confidently wrong

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py (hunk 3 above)
Comment (alejandromarcu) [reply deep_quoridor/src/agents/alphazero/nn_evaluator.py:54]: Maybe we should be using a logging tool with different levels of log that we can turn on and off, and this should be a warning

Comment (jonbinney) [reply deep_quoridor/src/agents/alphazero/nn_evaluator.py:54]: I think that once the policy is all zeros, you're completely hosed, right? This could raise an exception.

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py (hunk 1 above)
Comment (adamantivm) [reply deep_quoridor/src/agents/alphazero/nn_evaluator.py:57]: :thinking:

File: deep_quoridor/src/renderers/computation_times.py (hunk 4 above)
Comment (adamantivm) [reply deep_quoridor/src/renderers/computation_times.py:34]: :handshake:

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py (hunk 3 above)
Comment (adamantivm) [reply deep_quoridor/src/agents/alphazero/nn_evaluator.py:54]: I'll leave the check (after the softmax now) anyway, JIC, but I posit it will never trigger.

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py (hunk 1 above)
Comment (jonbinney) [reply deep_quoridor/src/agents/alphazero/nn_evaluator.py:57]: ok i think i may have found it! For the policy loss, they use  [tf.nn.softmax_cross_entropy_with_logits_v2(
            logits=policy_logits, labels=policy_targets), ](https://github.com/google-deepmind/open_spiel/blob/b11f9eb62a67a0982c0183d7170be6411c7447ab/open_spiel/python/algorithms/alpha_zero/model.py#L283), which knows that it is comparing log-probabilities with probabilities. We just use `cross_entropy()` loss, which expects both inputs to be probability distributions.

**Comment by adamantivm** [comment]
@jonbinney wdyt about merging this first and then working on the alleged missing mask rotation in a new PR? can I merge?

**Comment by jonbinney** [comment]
Merging sounds good!

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py [hunk 5]
```diff
@@ -41,21 +41,29 @@ def evaluate(self, game: Quoridor):
...
 
-        # Re-normalize probabilities after masking.
-        policy_probs = policy_masked
-        policy_probs = policy_probs / policy_probs.sum()
-
-        if not (np.isfinite(policy_probs).all() and np.isfinite(value)):
-            raise ValueError("Non-finite number in policy")
-
-        return value, policy_probs
+        if np.all(policy_masked == 0):
```
[the remark this answers is not in the document]
Comment (adamantivm) [reply deep_quoridor/src/agents/alphazero/nn_evaluator.py:52]: why do you say it will fail the assert below? the content inside the `if` block will make it so that it's no longer all zeroes. Can you clarify @alejandromarcu ?
